types/wlr_output_power_management_v1.c

```diff
@@ -0,0 +1,244 @@
+#include <assert.h>
+#include <fcntl.h>
+#include <stdlib.h>
+#include <unistd.h>
+#include <wayland-server-core.h>
+#include <wlr/interfaces/wlr_output.h>
+#include <wlr/types/wlr_output_power_management_v1.h>
+#include <wlr/types/wlr_output.h>
+#include <wlr/util/log.h>
+#include "util/signal.h"
+#include "wlr-output-power-management-unstable-v1-protocol.h"
+
+#define OUTPUT_POWER_MANAGER_V1_VERSION 1
+
+static void output_power_handle_destroy(struct wl_client *client,
+		struct wl_resource *resource) {
+	wl_resource_destroy(resource);
+}
+
+static void output_power_destroy(struct wlr_output_power_v1 *output_power) {
+	if (output_power == NULL) {
+		return;
+	}
+	wl_resource_set_user_data(output_power->resource, NULL);
+	wl_list_remove(&output_power->output_destroy_listener.link);
+	wl_list_remove(&output_power->output_enable_listener.link);
+	wl_list_remove(&output_power->link);
+	free(output_power);
+}
+
+static const struct zwlr_output_power_v1_interface output_power_impl;
+
+static struct wlr_output_power_v1 *output_power_from_resource(
+		struct wl_resource *resource) {
+	assert(wl_resource_instance_of(resource, &zwlr_output_power_v1_interface,
+		&output_power_impl));
+	return wl_resource_get_user_data(resource);
+}
+
+static void output_power_handle_resource_destroy(struct wl_resource *resource) {
+	struct wlr_output_power_v1 *output_power =
+		output_power_from_resource(resource);
+	output_power_destroy(output_power);
+}
+
+static void output_power_handle_output_destroy(struct wl_listener *listener,
+		void *data) {
+	struct wlr_output_power_v1 *output_power =
+		wl_container_of(listener, output_power, output_destroy_listener);
+	output_power_destroy(output_power);
+}
+
+static void wlr_output_power_v1_send_mode(struct wlr_output_power_v1 *output_power) {
+	enum zwlr_output_power_v1_mode mode;
+
+	mode = output_power->output->enabled ? ZWLR_OUTPUT_POWER_V1_MODE_ON :
+		ZWLR_OUTPUT_POWER_V1_MODE_OFF;
+	zwlr_output_power_v1_send_mode(output_power->resource, mode);
+}
+
+static void output_power_handle_output_enable(struct wl_listener *listener,
+		void *data) {
+	struct wlr_output_power_v1 *output_power =
+		wl_container_of(listener, output_power, output_enable_listener);
+	wlr_output_power_v1_send_mode(output_power);
+}
+
+static void output_power_handle_set_mode(struct wl_client *client,
+		struct wl_resource *output_power_resource,
+		enum zwlr_output_power_v1_mode mode) {
+	struct wlr_output_power_v1 *output_power =
+		output_power_from_resource(output_power_resource);
+	if (output_power == NULL) {
+		return;
+	}
+
+	switch (mode) {
+	case ZWLR_OUTPUT_POWER_V1_MODE_OFF:
+	case ZWLR_OUTPUT_POWER_V1_MODE_ON:
+		break;
+	default:
+		wlr_log(WLR_ERROR, "Invalid power mode %d", mode);
+		zwlr_output_power_v1_send_failed(output_power->resource);
```

Comment:
We should add a `return` right after to prevent `wlr_output_enable` from being called.

But actually sending an invalid enum value should be a protocol error (not because the output doesn't support it, but because it's an invalid value).

Reply:
i've introduced an error for that and updated the protocol